src/ray/core_worker/task_manager.cc

```diff
@@ -456,22 +462,35 @@ bool TaskManager::RetryTaskIfPossible(const TaskID &task_id) {
         << "Tried to retry task that was not pending " << task_id;
     spec = it->second.spec;
     num_retries_left = it->second.num_retries_left;
-    if (num_retries_left > 0) {
-      it->second.num_retries_left--;
+    num_oom_retries_left = it->second.num_oom_retries_left;
+    if (task_failed_due_to_oom) {
+      if (num_oom_retries_left > 0) {
+        will_retry = true;
+        it->second.num_oom_retries_left--;
+      }
     } else {
-      RAY_CHECK(num_retries_left == 0 || num_retries_left == -1);
+      if (num_retries_left > 0) {
+        will_retry = true;
+        it->second.num_retries_left--;
+      } else if (num_retries_left == -1) {
+        will_retry = true;
+      } else {
+        RAY_CHECK(num_retries_left == 0);
+      }
     }
     it->second.SetStatus(rpc::TaskStatus::SCHEDULED);
```

Comment:
if we decide to not retry, should the status still be SCHEDULED?

Reply:
I was wondering that too, I'll add a todo to look into in a follow up. If we change this now it might have side effects / break something

Reply:
assuming we are going to update this.

Reply:
Yes here's the new logic:

    if (task_failed_due_to_oom) {
      if (num_oom_retries_left > 0) {
        will_retry = true;
        it->second.num_oom_retries_left--;
      }
    } else {
      if (num_retries_left > 0) {
        will_retry = true;
        it->second.num_retries_left--;
      } else if (num_retries_left == -1) {
        will_retry = true;
      } else {
        RAY_CHECK(num_retries_left == 0);
      }
    }
    if (will_retry) {
      it->second.SetStatus(rpc::TaskStatus::SCHEDULED);
    }